File: frontend/src/features/blog/http/api/CreateBlog.ts
import type { BlogSummary } from '@/features/blog/http/response/BlogSummary'
import type { CreateBlogRequest } from '@/features/blog/http/request/CreateBlogRequest'
import {
  fromBlogSummaryContract,
  toCreateBlogRequestContract,
} from '@/features/blog/http/codec/BlogHttpCodecs'
import { postJson } from '@/shared/api/http-client'

export async function createBlog(request: CreateBlogRequest): Promise<BlogSummary> {
  return postJson('/api/blogs', fromBlogSummaryContract, toCreateBlogRequestContract(request))
}

File: backend/src/main/scala/domains/blog/http/api/CreateBlog.scala
package domains.blog.http.api



import domains.blog.http.*
import domains.blog.http.codec.BlogHttpCodecs.given
import cats.effect.IO
import domains.blog.application.BlogCommands
import domains.blog.model.request.CreateBlogRequest
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.Http4sDsl
import org.http4s.dsl.io.*

object CreateBlog:

  def routes(context: BlogHttpRouteContext)(using Http4sDsl[IO]): HttpRoutes[IO] =
    HttpRoutes.of[IO] {
      case request @ POST -> Root / "api" / "blogs" =>
        context.handlers.executeDecoded[CreateBlogRequest, BlogCommands.CreateBlogResult](
          request,
          context.plans.createBlog
        )
    }
